### PR TITLE
[PREGEL]Pregel SLLP

### PR DESCRIPTION
**Pull request by romanatarango**
### Scope & Purpose

Added a comment on the SLPA algorithm. The code didn't change, no tests are necessary.

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Pregel/Algos/SLPA.h [hunk 1]
```diff
@@ -40,6 +40,56 @@ namespace algos {
...
+ * described generally in the paper is based on the ArangoDB's documentation.
+ *
+ * The algorithm is performed in a series of iterations during that vertices
+ * send vertex IDs (natural numbers) to each other. Each vertex contains all
+ * vertex IDs it has received since the beginning together with the information
+ * how often an ID has been received. At the beginning, a vertex contains only
+ * its own ID that was "obtained" once. In an iteration, the vertices are
+ * processed in a random order, which is chosen for each iteration separately.
+ * A vertex v gets from each of its in-neighbors one of their saved IDs. Which
+ * if the saved IDs is sent by an in-neighbor n is chosen randomly with the
```
Comment (jvolmer) [inline arangod/Pregel/Algos/SLPA.h:60]:
```suggestion
 * of the saved IDs is sent by an in-neighbor n is chosen randomly with the
```

File: arangod/Pregel/Algos/SLPA.h [hunk 2]
```diff
@@ -40,6 +40,56 @@ namespace algos {
...
+ * probability
+ *
+ * <number of times the ID was obtained by n> / <number of times any ID
+ * was obtained by n>.
+ *
+ * Vertex n chooses one of the received IDs to save in the
```
Comment (jvolmer) [inline arangod/Pregel/Algos/SLPA.h:66]: I'm confused with this n. Before, n was the sending vertex. Perhaps better to use v here?

File: arangod/Pregel/Algos/SLPA.h [hunk 3]
```diff
@@ -40,6 +40,56 @@ namespace algos {
...
+ * iteration: it takes the least ID out of those that arrived most often. For
+ * example, if v obtained (3, 3, 2, 4, 4), it chooses 3.
```
Comment (jvolmer) [inline arangod/Pregel/Algos/SLPA.h:68]: Do  I understand correctly that in one iteration a vertex receives several numbers from its in-neighbors, only saves one of them and discards all the others?
I'm a bit confused with that: Is then the number of obtained IDs changed by one (the saved vertex) or by the number of all incoming numbers?

File: arangod/Pregel/Algos/SLPA.h [hunk 4]
```diff
@@ -40,6 +40,56 @@ namespace algos {
...
+ * correlates between different vertices. (Btw, the messages are sent anyway,
+ * but if the outcome is no, they are not processed.)
+ *
+ * (2) A vertex always sends its own ID (why?!) and sometimes a saved ID that is
+ * chosen as follows. Generate a random number r between 0 and the number of
+ * times the vertex received any IDs (i.e., the sum over all IDs of the numbers
+ * of times an ID was received). Now iterate (always in the same order) over the
+ * saved IDs of the vertex and trace the sum of times they were received. When
+ * Thus accumulated sum reaches r (the randomly generated number), the the
+ * current ID and send it.
```
Comment (jvolmer) [inline arangod/Pregel/Algos/SLPA.h:91]:
```suggestion
 * saved IDs of the vertex and trace the sum of times they were received. When
 * thus accumulated sum reaches r (the randomly generated number), choose the
 * current ID and send it.
```

**Review by jvolmer — COMMENTED** [review]
Good idea to write down how the algorithm works / is supposed to work. Two things confused me a bit, perhaps you can make these a bit clearer?

File: arangod/Pregel/Algos/SLPA.h (hunk 2 above)
Comment (romanatarango) [reply arangod/Pregel/Algos/SLPA.h:66]: It is a _n_eighbor of v. Would w be better?

Comment (romanatarango) [reply arangod/Pregel/Algos/SLPA.h]: I applied it, please, check along with the other changes.

File: arangod/Pregel/Algos/SLPA.h (hunk 3 above)
Comment (romanatarango) [reply arangod/Pregel/Algos/SLPA.h]: Yes, you understand it correctly. The number is changed by one. Adapted the description.

**Review by jvolmer — APPROVED** [review]

**Pull request merged**